deepspeed/runtime/domino/transformer.py

```diff
@@ -260,21 +260,47 @@ def __init__(self,
                                                 no_persist_layer_norm=config.no_persist_layer_norm)
 
         # Self attention.
-        self.self_attention = ShardedAttention(config,
-                                               layer_number,
-                                               mpu,
-                                               ColumnParallelLinear,
-                                               RowParallelLinearNoComm,
-                                               apply_rotary_pos_emb,
-                                               attention_type=AttnType.self_attn,
-                                               attn_mask_type=self_attn_mask_type)
+        if self.input_split_dim == "batch":
+            self.self_attention = ShardedAttention(config,
+                                                layer_number,
+                                                mpu,
+                                                ColumnParallelLinear,
+                                                RowParallelLinearNoComm,
+                                                apply_rotary_pos_emb,
+                                                attention_type=AttnType.self_attn,
+                                                attn_mask_type=self_attn_mask_type)
+        elif self.input_split_dim == "seq":
+            query_projection_size = config.kv_channels * config.num_attention_heads
+            kv_projection_size = config.kv_channels * config.num_attention_heads
+
+            # Per attention head and per partition values.
+            world_size = mpu.get_tensor_model_parallel_world_size()
+            self.hidden_size_per_attention_head = query_projection_size // config.num_attention_heads
+            self.num_attention_heads_per_partition = config.num_attention_heads // world_size
+            self.query_key_value = ColumnParallelLinear(config.hidden_size,
+                                                    query_projection_size + 2 * kv_projection_size,
+                                                    config=config,
+                                                    init_method=config.init_method,
+                                                    bias=config.add_bias_linear,
+                                                    gather_output=False)
+            self.self_attention_sp = CoreAttention(config, self.layer_number, mpu, self_attn_mask_type)
+            self.dense = RowParallelLinearNoComm(query_projection_size,
+                                                config.hidden_size,
+                                                config=config,
+                                                init_method=config.output_layer_init_method,
+                                                bias=config.add_bias_linear,
+                                                input_is_parallel=True,
+                                                skip_bias_add=True)
```

Comment:
these lines are mostly identical to our shardedAttention module, even the new component `self.self_attention_sp` initialized the same way as original `self.core_attention`. I don't see why we need this duplication.